bgpd/bgp_addpath.c

```diff
@@ -0,0 +1,410 @@
+	bgp->tx_addpath.id_allocators[afi][safi][addpath_type] = NULL;
+}
+
+/*
+ * Allocate an Addpath ID for the given type on a path, if necessary.
+ */
+static void bgp_addpath_populate_path(struct id_alloc *allocator,
+				      struct bgp_info *path,
+				      bgp_addpath_strategy_e addpath_type)
+{
+	if (bgp_addpath_tx_path(addpath_type, path)) {
+		path->tx_addpath.addpath_tx_id[addpath_type] =
+			idalloc_allocate(allocator);
+	}
+}
+
+/*
+ * Compute addpath ID's on a BGP instance associated with the addpath strategy,
+ * and afi/safi combination. Since we won't waste the time computing addpath IDs
+ * for unused strategies, the first time a peer is configured to use a strategy,
+ * we have to backfill the data.
+ */
+static void bgp_addpath_populate_type(struct bgp *bgp, afi_t afi, safi_t safi,
+				    bgp_addpath_strategy_e addpath_type)
+{
+	struct bgp_node *rn;
+	struct bgp_info *bi;
+	char buf[200];
+	struct id_alloc *allocator;
+
+	snprintf(buf, sizeof(buf), "Addpath ID Allocator %s:%d/%d",
+		 bgp_addpath_names(addpath_type)->config_name, (int)afi,
+		 (int)safi);
+	buf[sizeof(buf) - 1] = '\0';
+	zlog_info("Computing addpath IDs for addpath type %s",
+		bgp_addpath_names(addpath_type)->human_name);
+
+	bgp->tx_addpath.id_allocators[afi][safi][addpath_type] =
+		idalloc_new(buf);
+
+	idalloc_reserve(bgp->tx_addpath.id_allocators[afi][safi][addpath_type],
+		BGP_ADDPATH_TX_ID_FOR_DEFAULT_ORIGINATE);
+
+	allocator = bgp->tx_addpath.id_allocators[afi][safi][addpath_type];
+
+	for (rn = bgp_table_top(bgp->rib[afi][safi]); rn;
+	     rn = bgp_route_next(rn))
+		for (bi = rn->info; bi; bi = bi->next)
+			bgp_addpath_populate_path(allocator, bi, addpath_type);
+}
+
+/*
+ * Handle updates to a peer or group's addpath strategy. If after adjusting
+ * counts a addpath strategy is in use for the first time, or no longer in use,
+ * the IDs for that strategy will be populated or flushed.
+ */
+void bgp_addpath_type_changed(struct bgp *bgp, afi_t afi, safi_t safi,
+				   bgp_addpath_strategy_e old_strat,
+				   bgp_addpath_strategy_e new_strat)
+{
+	if (old_strat == new_strat)
+		return;
+
+	if (old_strat != BGP_ADDPATH_NONE) {
+		bgp->tx_addpath.peercount[afi][safi][old_strat] -= 1;
+		bgp->tx_addpath.total_peercount[afi][safi] -= 1;
+
+		if (bgp->tx_addpath.peercount[afi][safi][old_strat] == 0)
+			bgp_addpath_flush_type(bgp, afi, safi, old_strat);
+	}
+
+	if (new_strat != BGP_ADDPATH_NONE) {
+		if (bgp->tx_addpath.peercount[afi][safi][new_strat] == 0)
+			bgp_addpath_populate_type(bgp, afi, safi, new_strat);
+
+		bgp->tx_addpath.peercount[afi][safi][new_strat] += 1;
+		bgp->tx_addpath.total_peercount[afi][safi] += 1;
+	}
+}
+
+/*
+ * Change the addpath type assigned to a peer, or peer group. In addition to
+ * adjusting the counts, peer sessions will be reset as needed to make the
+ * change take effect.
+ */
+void peer_af_addpath_set_type(struct peer *peer, afi_t afi, safi_t safi,
```

Comment:
Change to bgp_addpath_set_type_for_peer or bgp_addpath_peer_af_set_type? Ok otherwise for me.